lib/Modules/FreeRdpHooks/FreeRdpHooks.psm1

```diff
@@ -30,11 +30,19 @@ function Get-Vcredist {
         try {
             Write-JujuWarning "Trying to get vcredist Juju resource"
             $vcredistPath = Get-JujuResource -Resource "vcredist-x64"
-            return $vcredistPath
         } catch {
+            $vcredistPath = $null
             Write-JujuWarning "Failed downloading vcredist resource: $_"
-            Write-JujuWarning "Falling back to file download"
         }
+        if ($vcredistPath) {
+            $bytes = Get-Content $vcredistPath -TotalCount 31 -Encoding Byte
+            $str = [string]::join("", [char[]]$bytes)
+            if ($str -eq $DEFAULT_JUJU_RESOURCE_CONTENT) {
+                Throw "Manually attach the Juju resource 'vcredist-x64' or set the config option 'vcredist-url'."
```

Comment:
If I am reading this correctly, you are trying to download it from the URL in the config, if no URL is set, you attempt to download the resource from the controller, then you throw. My point is, that missing resources should not error out the charm. At least not without a clear message of what is wrong set in the status (Which I don't see happening here).

This is a configuration issue. Errors should be returned only when the charm has encountered a condition (no network connectivity, not enough disk space, read-only file systems, incorrect parameters passed to installer, etc) from which it cannot recover in any way without user/developer intervention. Information that you expect the user to fill in, but you cannot find, should block , not error.

You can finish any installation in any of the other hooks. FreeRDP does not have a huge installation footprint. You can add some idempotency here, and call install when conditions are met, block and return otherwise.